// enzyme-adapter-react-helper/build/safeSFC.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _react = _interopRequireDefault(require("react"));

var _object = _interopRequireDefault(require("object.getownpropertydescriptors"));

var _ifReact = _interopRequireDefault(require("./ifReact"));

function _createSuper(Derived) { return function () { var Super = (0, _getPrototypeOf2["default"])(Derived), result; if (_isNativeReflectConstruct()) { var NewTarget = (0, _getPrototypeOf2["default"])(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2["default"])(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function assertFunction(fn) {
  if (typeof fn !== 'function') {
    throw new TypeError('Component must be a function');
  }

  return fn;
}

function copyStatics(source, target) {
  assertFunction(source); // eslint-disable-next-line no-param-reassign

  target.displayName = source.displayName || source.name;

  var _gOPDs = (0, _object["default"])(source),
      oldProto = _gOPDs.prototype,
      descriptors = (0, _objectWithoutProperties2["default"])(_gOPDs, ["prototype"]);

  Object.defineProperties(target, descriptors);
  return target;
}

function nullToNoScript(fn) {
  // eslint-disable-next-line prefer-arrow-callback
  return copyStatics(fn, function NullHandler() {
    var element = fn.apply(void 0, arguments);
    return element === null ? /*#__PURE__*/_react["default"].createElement("noscript", null) : element;
  });
}

var maybeNullWrapper = (0, _ifReact["default"])('^0.14', nullToNoScript, assertFunction);

function safeSFC(fn) {
  return copyStatics(fn, /*#__PURE__*/function (_React$Component) {
    (0, _inherits2["default"])(SafeSFC, _React$Component);

    var _super = _createSuper(SafeSFC);

    function SafeSFC() {
      (0, _classCallCheck2["default"])(this, SafeSFC);
      return _super.apply(this, arguments);
    }

    (0, _createClass2["default"])(SafeSFC, [{
      key: "render",
      value: function render() {
        return maybeNullWrapper(fn(this.props, this.context));
      }
    }]);
    return SafeSFC;
  }(_react["default"].Component));
}

var _default = (0, _ifReact["default"])('>= 0.14', maybeNullWrapper, safeSFC);

exports["default"] = _default;
module.exports = exports.default;
